# ğŸ”¢ API VERSIONING - OBRIGATÃ“RIO

**VersÃ£o**: 1.0  
**Data**: 05/Dez/2025  
**Prioridade**: CRÃTICA

---

## ğŸ¯ REGRA DE OURO

**TODAS as APIs REST devem ter versionamento explÃ­cito no path.**

---

## âœ… PADRÃƒO OBRIGATÃ“RIO

### Estrutura de URLs
```
/{app}/api/v1/{resource}/
```

### Exemplos Corretos
```python
# âœ… CORRETO
/kpi/api/v1/uploads/
/kpi/api/v1/uploads/{id}/
/dashboard/api/v1/stats/
/dashboard/api/v1/charts/
/users/api/v1/profile/
```

### Exemplos Errados
```python
# âŒ ERRADO - Sem versionamento
/api/uploads/
/kpi/api/uploads/
/dashboard/stats/

# âŒ ERRADO - Versionamento no header
Headers: { 'API-Version': 'v1' }

# âŒ ERRADO - Versionamento no query param
/api/uploads/?version=v1
```

---

## ğŸ“ ESTRUTURA DE ARQUIVOS

```
app/
â”œâ”€â”€ urls.py
â”‚   â””â”€â”€ urlpatterns = [
â”‚         path('', TemplateView),           # Template
â”‚         path('api/v1/', include('app.api.urls')),  # API v1
â”‚       ]
â”œâ”€â”€ api/
â”‚   â”œâ”€â”€ __init__.py
â”‚   â”œâ”€â”€ urls.py                 # URLs da API
â”‚   â”‚   â””â”€â”€ urlpatterns = [
â”‚   â”‚         path('resource/', ResourceListView),
â”‚   â”‚         path('resource/<id>/', ResourceDetailView),
â”‚   â”‚       ]
â”‚   â”œâ”€â”€ views.py                # API Views
â”‚   â”œâ”€â”€ serializers.py          # Serializers DRF
â”‚   â””â”€â”€ validators.py           # Validators
```

---

## ğŸ”§ IMPLEMENTAÃ‡ÃƒO

### URLs Principais
```python
# app/urls.py
from django.urls import path, include

urlpatterns = [
    # Template views (sem /api/)
    path('', DashboardView.as_view(), name='dashboard'),
    path('upload/', UploadView.as_view(), name='upload'),
    
    # API routes (com versionamento)
    path('api/v1/', include('app.api.urls')),
]
```

### URLs da API
```python
# app/api/urls.py
from django.urls import path
from .views import (
    FileUploadListAPIView,
    FileUploadDetailAPIView,
    FileUploadCreateAPIView,
)

urlpatterns = [
    path('uploads/', FileUploadListAPIView.as_view(), name='api_uploads_list'),
    path('uploads/<uuid:id>/', FileUploadDetailAPIView.as_view(), name='api_uploads_detail'),
    path('uploads/create/', FileUploadCreateAPIView.as_view(), name='api_uploads_create'),
]
```

### API Views
```python
# app/api/views.py
from rest_framework.views import APIView
from rest_framework.response import Response
from rest_framework import status

from .serializers import FileUploadSerializer
from ..controllers import FileUploadController

class FileUploadCreateAPIView(APIView):
    """
    POST /app/api/v1/uploads/create/
    Create new file upload
    """
    
    def post(self, request):
        file_obj = request.FILES.get('file')
        
        result = FileUploadController.create_upload(file_obj)
        
        if not result['success']:
            return Response(
                {'error': result['error']},
                status=status.HTTP_400_BAD_REQUEST
            )
        
        serializer = FileUploadSerializer(result['upload'])
        return Response(serializer.data, status=status.HTTP_201_CREATED)
```

---

## ğŸ”„ MIGRAÃ‡ÃƒO DE VERSÃ•ES

### Quando Criar Nova VersÃ£o?

Crie `v2` quando:
- Breaking changes na API
- MudanÃ§a de estrutura de response
- RemoÃ§Ã£o de campos
- MudanÃ§a de tipos de dados

### Como Manter MÃºltiplas VersÃµes
```python
app/
â”œâ”€â”€ api/
â”‚   â”œâ”€â”€ v1/
â”‚   â”‚   â”œâ”€â”€ urls.py
â”‚   â”‚   â”œâ”€â”€ views.py
â”‚   â”‚   â””â”€â”€ serializers.py
â”‚   â””â”€â”€ v2/
â”‚       â”œâ”€â”€ urls.py
â”‚       â”œâ”€â”€ views.py
â”‚       â””â”€â”€ serializers.py
```

---

## ğŸ“Š VERSIONAMENTO NO SISTEMA

### Config URLs
```python
# config/urls.py
urlpatterns = [
    path('kpi/', include('kpi.urls')),
    path('dashboard/', include('dashboard.urls')),
]
```

### App URLs
```python
# kpi/urls.py
urlpatterns = [
    # Templates
    path('upload/', UploadView.as_view()),
    
    # API v1
    path('api/v1/', include('kpi.api.urls')),
]
```

### Resultado Final
```
/kpi/upload/              â†’ Template
/kpi/api/v1/uploads/      â†’ API v1
/dashboard/               â†’ Template
/dashboard/api/v1/stats/  â†’ API v1
```

---

## ğŸš« VIOLAÃ‡Ã•ES COMUNS

### âŒ Exemplo 1: API sem VersÃ£o
```python
# âŒ ERRADO
path('api/uploads/', UploadAPIView)

# âœ… CORRETO
path('api/v1/uploads/', UploadAPIView)
```

### âŒ Exemplo 2: VersÃ£o no Nome
```python
# âŒ ERRADO
path('api/uploads/', UploadV1APIView)  # VersÃ£o no nome da classe

# âœ… CORRETO
path('api/v1/uploads/', UploadAPIView)  # VersÃ£o no path
```

### âŒ Exemplo 3: Misturar Templates e API
```python
# âŒ ERRADO
urlpatterns = [
    path('uploads/', UploadView),        # Template
    path('uploads/create/', UploadAPI),  # API sem /api/v1/
]

# âœ… CORRETO
urlpatterns = [
    path('upload/', UploadView),              # Template
    path('api/v1/uploads/', UploadListAPI),   # API
    path('api/v1/uploads/create/', UploadCreateAPI),
]
```

---

## ğŸ“ DOCUMENTAÃ‡ÃƒO DE API

### Swagger/OpenAPI
Todas as APIs devem ser documentadas:
```python
from drf_yasg.views import get_schema_view
from drf_yasg import openapi

schema_view = get_schema_view(
    openapi.Info(
        title="KPI API",
        default_version='v1',
    ),
)

urlpatterns = [
    path('api/v1/docs/', schema_view.with_ui('swagger')),
]
```

---

## âœ… CHECKLIST

### Estrutura
- [ ] APIs em `app/api/`
- [ ] Templates em `app/views.py`
- [ ] Versionamento `api/v1/`
- [ ] Serializers criados
- [ ] Controllers existem
- [ ] Repositories existem

### URLs
- [ ] Templates: `/{app}/{page}/`
- [ ] APIs: `/{app}/api/v1/{resource}/`
- [ ] Sem mistura de concerns

### CÃ³digo
- [ ] Views de template sem lÃ³gica
- [ ] API views usam controllers
- [ ] Serializers para todos os responses
- [ ] Sem ORM direto nas views

---

**API = /api/v1/. TEMPLATES = SEM /api/.**

