# üì§ RETORNOS: Controllers e APIs

**Vers√£o**: 1.0  
**Data**: 05/Dez/2025  
**Prioridade**: CR√çTICA

---

## üéØ REGRA DE OURO

**Controllers retornam OBJETOS. APIs serializam e definem HTTP status.**

---

## ‚úÖ PADR√ÉO OBRIGAT√ìRIO

### Controllers (Business Logic)

**INSERT/CREATE**:
```python
# ‚úÖ CORRETO - Retorna objeto ou None
def create_upload(file_obj):
    validation = Validator.validate(file_obj)
    if not validation['valid']:
        return None  # Ou raise ValidationError
    
    upload = Repository.create(file=file_obj)
    return upload  # Retorna o objeto criado
```

**UPDATE**:
```python
# ‚úÖ CORRETO - Retorna objeto atualizado
def update_status(upload_id, new_status):
    upload = Repository.get_by_id(upload_id)
    if not upload:
        return None
    
    upload.status = new_status
    Repository.save(upload)
    return upload  # Retorna o objeto atualizado
```

**DELETE**:
```python
# ‚úÖ CORRETO - Retorna None ou booleano
def delete_upload(upload_id):
    upload = Repository.get_by_id(upload_id)
    if not upload:
        return False
    
    Repository.delete(upload)
    return True  # Ou None
```

**LIST**:
```python
# ‚úÖ CORRETO - Retorna QuerySet ou lista de objetos
def list_uploads(filters):
    return Repository.filter(**filters)  # QuerySet
```

**RETRIEVE**:
```python
# ‚úÖ CORRETO - Retorna objeto ou None
def get_upload(upload_id):
    return Repository.get_by_id(upload_id)  # Objeto ou None
```

---

## üîå API Views (Serializa√ß√£o e Status)

### POST (CREATE) - 201 Created
```python
class CreateAPIView(APIView):
    def post(self, request):
        file_obj = request.FILES.get('file')
        
        # Controller retorna objeto ou None
        upload = FileUploadController.create_upload(file_obj)
        
        if not upload:
            return Response(
                {'error': 'Validation failed'},
                status=status.HTTP_400_BAD_REQUEST
            )
        
        # Serializar objeto
        serializer = FileUploadSerializer(upload, context={'request': request})
        return Response(serializer.data, status=status.HTTP_201_CREATED)
```

### GET (LIST) - 200 OK
```python
class ListAPIView(APIView):
    def get(self, request):
        # Controller retorna QuerySet ou lista
        uploads = FileUploadController.list_uploads(request.GET.dict())
        
        # Serializar
        serializer = FileUploadSerializer(uploads, many=True, context={'request': request})
        return Response(serializer.data, status=status.HTTP_200_OK)
```

### GET (RETRIEVE) - 200 OK ou 404 Not Found
```python
class DetailAPIView(APIView):
    def get(self, request, upload_id):
        # Controller retorna objeto ou None
        upload = FileUploadController.get_upload(upload_id)
        
        if not upload:
            return Response(
                {'error': 'Not found'},
                status=status.HTTP_404_NOT_FOUND
            )
        
        # Serializar
        serializer = FileUploadSerializer(upload, context={'request': request})
        return Response(serializer.data, status=status.HTTP_200_OK)
```

### PUT/PATCH (UPDATE) - 200 OK ou 404 Not Found
```python
class UpdateAPIView(APIView):
    def patch(self, request, upload_id):
        # Controller retorna objeto atualizado ou None
        upload = FileUploadController.update_upload(upload_id, request.data)
        
        if not upload:
            return Response(
                {'error': 'Not found or validation failed'},
                status=status.HTTP_404_NOT_FOUND
            )
        
        # Serializar
        serializer = FileUploadSerializer(upload, context={'request': request})
        return Response(serializer.data, status=status.HTTP_200_OK)
```

### DELETE - 204 No Content ou 404 Not Found
```python
class DeleteAPIView(APIView):
    def delete(self, request, upload_id):
        # Controller retorna True/False ou None
        deleted = FileUploadController.delete_upload(upload_id)
        
        if not deleted:
            return Response(
                {'error': 'Not found'},
                status=status.HTTP_404_NOT_FOUND
            )
        
        # Sem conte√∫do
        return Response(status=status.HTTP_204_NO_CONTENT)
```

---

## üö´ VIOLA√á√ïES COMUNS

### ‚ùå Exemplo 1: Controller com Dict de Status
```python
# ‚ùå ERRADO
def create_upload(file_obj):
    upload = Repository.create(file=file_obj)
    return {
        'success': True,  # ‚ùå N√£o precisa
        'message': 'Created',  # ‚ùå N√£o precisa
        'upload': upload,
        'error': None  # ‚ùå N√£o precisa
    }

# ‚úÖ CORRETO
def create_upload(file_obj):
    validation = Validator.validate(file_obj)
    if not validation['valid']:
        return None  # Ou raise exception
    
    upload = Repository.create(file=file_obj)
    return upload  # Apenas o objeto
```

### ‚ùå Exemplo 2: API sem Serializer
```python
# ‚ùå ERRADO
def post(self, request):
    upload = Controller.create_upload(file)
    return Response({
        'id': upload.id,  # Serializa√ß√£o manual
        'status': upload.status
    }, status=201)

# ‚úÖ CORRETO
def post(self, request):
    upload = Controller.create_upload(file)
    if not upload:
        return Response({'error': 'Failed'}, status=400)
    
    serializer = FileUploadSerializer(upload, context={'request': request})
    return Response(serializer.data, status=201)
```

### ‚ùå Exemplo 3: Controller Define HTTP Status
```python
# ‚ùå ERRADO
def create_upload(file_obj):
    if not file_obj:
        return {'status': 400, 'error': 'No file'}  # ‚ùå
    
    upload = Repository.create(file=file_obj)
    return {'status': 201, 'data': upload}  # ‚ùå

# ‚úÖ CORRETO
def create_upload(file_obj):
    if not file_obj:
        return None  # API decide status
    
    upload = Repository.create(file=file_obj)
    return upload  # API decide status
```

---

## üìã TABELA DE RETORNOS

| Opera√ß√£o | Controller Retorna | API Status | API Response |
|----------|-------------------|------------|--------------|
| **CREATE** | Objeto ou None | 201 Created ou 400 Bad Request | Serializer(objeto).data |
| **UPDATE** | Objeto ou None | 200 OK ou 404 Not Found | Serializer(objeto).data |
| **DELETE** | True/False/None | 204 No Content ou 404 Not Found | Vazio |
| **LIST** | QuerySet/List | 200 OK | Serializer(lista, many=True).data |
| **RETRIEVE** | Objeto ou None | 200 OK ou 404 Not Found | Serializer(objeto).data |

---

## ‚úÖ CHECKLIST

### Controller
- [ ] Retorna objeto (n√£o dict com 'success')
- [ ] Retorna None em caso de falha (n√£o dict com 'error')
- [ ] N√£o define HTTP status
- [ ] N√£o cria mensagens de resposta
- [ ] Foca em l√≥gica de neg√≥cio

### API View
- [ ] Verifica se controller retornou None
- [ ] Define HTTP status apropriado
- [ ] Usa serializer para resposta
- [ ] Cria mensagens de erro quando necess√°rio
- [ ] N√£o faz l√≥gica de neg√≥cio

---

## üîÑ TRATAMENTO DE ERROS

### Controller - Exceptions
```python
def create_upload(file_obj):
    validation = Validator.validate(file_obj)
    if not validation['valid']:
        # Op√ß√£o 1: Retornar None
        return None
        
        # Op√ß√£o 2: Raise exception (prefer√≠vel)
        raise ValidationError(validation['error'])
    
    upload = Repository.create(file=file_obj)
    return upload
```

### API View - HTTP Status
```python
def post(self, request):
    try:
        upload = Controller.create_upload(file)
        
        if not upload:
            return Response({'error': 'Validation failed'}, status=400)
        
        serializer = Serializer(upload, context={'request': request})
        return Response(serializer.data, status=201)
        
    except ValidationError as e:
        return Response({'error': str(e)}, status=400)
    except Exception as e:
        logger.error(f"Unexpected error: {e}")
        return Response({'error': 'Internal error'}, status=500)
```

---

**CONTROLLERS = OBJETOS. APIs = SERIALIZA√á√ÉO + STATUS.**

