# âš–ï¸ SEPARAÃ‡ÃƒO: Views de Template vs Views de API

**VersÃ£o**: 1.0  
**Data**: 05/Dez/2025  
**Prioridade**: CRÃTICA

---

## ğŸ¯ REGRA DE OURO

**Views de Template** e **Views de API** devem estar em arquivos separados com responsabilidades claras.

---

## ğŸ“ ESTRUTURA OBRIGATÃ“RIA

```
app/
â”œâ”€â”€ views.py                    # Views de Template APENAS
â”œâ”€â”€ api/
â”‚   â”œâ”€â”€ __init__.py
â”‚   â”œâ”€â”€ views.py               # Views de API (endpoints REST)
â”‚   â”œâ”€â”€ serializers.py         # Serializers DRF
â”‚   â””â”€â”€ validators.py          # Validadores de entrada
â”œâ”€â”€ controllers/
â”‚   â””â”€â”€ {model}_controller.py  # Business logic
â”œâ”€â”€ repositories/
â”‚   â””â”€â”€ {model}_repository.py  # Data access
â””â”€â”€ urls.py                     # Rotas (templates + API)
```

---

## ğŸ“„ VIEWS DE TEMPLATE

### Responsabilidades
âœ… **Apenas servir templates HTML**  
âœ… **Passar dados bÃ¡sicos** (dropdowns, filtros)  
âœ… **AutenticaÃ§Ã£o/AutorizaÃ§Ã£o** (decorators/mixins)  
âœ… **Renderizar** template com context

âŒ **NUNCA** fazer processamento de dados  
âŒ **NUNCA** fazer queries complexas  
âŒ **NUNCA** fazer cÃ¡lculos  
âŒ **NUNCA** acessar ORM diretamente (usar repositories)

### Estrutura
```python
# app/views.py
from django.contrib.auth.mixins import LoginRequiredMixin
from django.views import View
from django.shortcuts import render

class DashboardView(LoginRequiredMixin, View):
    """Serve dashboard template with basic data."""
    
    def get(self, request):
        # Only basic data for dropdowns/filters
        context = {
            'years': [2024, 2025],
            'units': UnitRepository.get_all(),
        }
        return render(request, 'dashboard.html', context)
```

### O que NÃƒO fazer
```python
# âŒ ERRADO - Processamento na view
def get(self, request):
    # Processamento pesado
    data = compute_kpis()
    charts = generate_charts()
    return render(request, 'dashboard.html', {'data': data})

# âŒ ERRADO - ORM direto
def get(self, request):
    uploads = FileUpload.objects.filter(status='PENDING')  # âŒ
    return render(request, 'uploads.html', {'uploads': uploads})

# âŒ ERRADO - POST com lÃ³gica de negÃ³cio
def post(self, request):
    file = request.FILES['file']
    # ValidaÃ§Ã£o inline
    if not file.name.endswith('.xlsx'):
        return redirect('error')
    # Processamento inline
    process_file(file)
    return redirect('success')
```

---

## ğŸ”Œ VIEWS DE API

### Responsabilidades
âœ… **Receber dados** do request  
âœ… **Chamar controller** com dados  
âœ… **Serializar resposta**  
âœ… **Retornar JSON**  
âœ… **Tratar erros** (HTTP status codes)

âŒ **NUNCA** acessar ORM diretamente  
âŒ **NUNCA** fazer validaÃ§Ã£o inline  
âŒ **NUNCA** fazer lÃ³gica de negÃ³cio

### Estrutura
```python
# app/api/views.py
from rest_framework.views import APIView
from rest_framework.response import Response
from rest_framework import status

from .serializers import FileUploadSerializer
from ..controllers import FileUploadController

class FileUploadAPIView(APIView):
    """
    API endpoint for file uploads.
    POST /api/v1/uploads/ - Create upload
    """
    
    def post(self, request):
        # 1. Get data from request
        file_obj = request.FILES.get('file')
        
        # 2. Call controller
        result = FileUploadController.create_upload(file_obj)
        
        # 3. Handle result
        if not result['success']:
            return Response(
                {'error': result['error']},
                status=status.HTTP_400_BAD_REQUEST
            )
        
        # 4. Serialize and return
        serializer = FileUploadSerializer(result['upload'])
        return Response(serializer.data, status=status.HTTP_201_CREATED)
```

---

## ğŸ”„ FLUXO COMPLETO

### Template View (GET)
```
Request â†’ View â†’ Repository.get_basic_data() â†’ Context â†’ Render Template
```

### API View (POST)
```
Request â†’ API View â†’ Controller â†’ Validator â†’ Repository â†’ ORM
                  â†“
            Serializer â†’ Response JSON
```

### Frontend (JavaScript)
```
User Action â†’ Fetch API â†’ JSON â†’ Alpine.js â†’ Update DOM
```

---

## ğŸš« VIOLAÃ‡Ã•ES COMUNS

### âŒ Exemplo 1: View com LÃ³gica de NegÃ³cio
```python
# âŒ ERRADO
class UploadView(View):
    def post(self, request):
        file = request.FILES['file']
        
        # ValidaÃ§Ã£o inline âŒ
        if not file.name.endswith('.xlsx'):
            return HttpResponse('Invalid file', status=400)
        
        # ORM direto âŒ
        upload = FileUpload.objects.create(file=file)
        
        # Processamento âŒ
        process_excel(upload.file.path)
        
        return redirect('success')

# âœ… CORRETO
class UploadView(View):
    def get(self, request):
        context = {}
        return render(request, 'upload.html', context)
    
    # POST via API, nÃ£o via template view
```

### âŒ Exemplo 2: API sem Serializer
```python
# âŒ ERRADO
class UploadAPIView(APIView):
    def post(self, request):
        file = request.FILES['file']
        upload = FileUpload.objects.create(file=file)  # ORM direto âŒ
        
        return Response({
            'id': upload.id,  # SerializaÃ§Ã£o manual âŒ
            'status': upload.status
        })

# âœ… CORRETO
class UploadAPIView(APIView):
    def post(self, request):
        file = request.FILES.get('file')
        result = FileUploadController.create_upload(file)
        
        serializer = FileUploadSerializer(result['upload'])
        return Response(serializer.data, status=201)
```

### âŒ Exemplo 3: View com Filtros Complexos
```python
# âŒ ERRADO
class ListUploadsView(View):
    def get(self, request):
        # Filtros complexos na view âŒ
        uploads = FileUpload.objects.filter(status='COMPLETED')
        if request.GET.get('month'):
            uploads = uploads.filter(uploaded_at__month=...)
        
        context = {'uploads': uploads}
        return render(request, 'list.html', context)

# âœ… CORRETO
class ListUploadsView(View):
    def get(self, request):
        # Apenas serve template, dados via API
        return render(request, 'list.html', {})

# API separada
class ListUploadsAPIView(APIView):
    def get(self, request):
        filters = request.GET.dict()
        result = FileUploadController.list_uploads(filters)
        serializer = FileUploadSerializer(result, many=True)
        return Response(serializer.data)
```

---

## âœ… CHECKLIST

### Antes de Criar View

- [ ] Ã‰ template ou API?
- [ ] Template: Apenas GET com dados bÃ¡sicos?
- [ ] API: Usa controller + serializer?
- [ ] NÃ£o tem ORM direto?
- [ ] NÃ£o tem lÃ³gica de negÃ³cio?

### Estrutura de Arquivos

- [ ] `views.py` - Apenas templates
- [ ] `api/views.py` - Apenas APIs REST
- [ ] `api/serializers.py` - Existe
- [ ] `controllers/` - Existe
- [ ] `repositories/` - Existe
- [ ] URLs separadas (templates vs API)

---

## ğŸ“š REFERÃŠNCIAS

- Controller Pattern: `CONTROLLER_PATTERN_OBRIGATORIO.md`
- Repository Pattern: `REPOSITORY_PATTERN_OBRIGATORIO.md`
- API Versioning: `API_VERSIONING_OBRIGATORIO.md`

---

**VIEWS = THIN LAYER. CONTROLLERS = BUSINESS LOGIC.**

