# ‚úÖ VALIDATORS - CAMADA OBRIGAT√ìRIA

**Vers√£o**: 1.0  
**Data**: 05/Dez/2025  
**Prioridade**: CR√çTICA

---

## üéØ REGRA DE OURO

**Controllers NUNCA fazem valida√ß√£o inline. SEMPRE usam classes Validator.**

---

## üìÅ ESTRUTURA OBRIGAT√ìRIA

```
app/
‚îú‚îÄ‚îÄ api/
‚îÇ   ‚îî‚îÄ‚îÄ validators.py          # Validators para APIs
‚îú‚îÄ‚îÄ controllers/
‚îÇ   ‚îî‚îÄ‚îÄ {model}_controller.py  # Usa validators
‚îî‚îÄ‚îÄ repositories/
    ‚îî‚îÄ‚îÄ {model}_repository.py  # N√£o valida, apenas persiste
```

---

## üìù TEMPLATE DE VALIDATOR

```python
# app/api/validators.py

class FileUploadValidator:
    """Validates file upload data."""
    
    @staticmethod
    def validate_file(file_obj):
        """
        Validate uploaded file.
        
        Args:
            file_obj: Django UploadedFile object
            
        Returns:
            dict: {'valid': bool, 'error': str|None, 'data': dict}
        """
        errors = []
        
        # Check if file exists
        if not file_obj:
            return {
                'valid': False,
                'error': 'No file provided',
                'data': None
            }
        
        # Check file extension
        allowed_extensions = ['.xlsx', '.xls']
        file_ext = file_obj.name.split('.')[-1].lower()
        
        if f'.{file_ext}' not in allowed_extensions:
            errors.append(f'Invalid file type. Allowed: {", ".join(allowed_extensions)}')
        
        # Check file size (max 50MB)
        max_size = 50 * 1024 * 1024  # 50MB in bytes
        if file_obj.size > max_size:
            errors.append(f'File too large. Max size: 50MB')
        
        # Check if file is readable
        try:
            file_obj.seek(0)
        except Exception as e:
            errors.append(f'Cannot read file: {str(e)}')
        
        if errors:
            return {
                'valid': False,
                'error': '; '.join(errors),
                'data': None
            }
        
        return {
            'valid': True,
            'error': None,
            'data': {
                'filename': file_obj.name,
                'size': file_obj.size,
                'content_type': file_obj.content_type
            }
        }
```

---

## üéØ RESPONSABILIDADES

### Validators
‚úÖ **Validar tipos** de dados  
‚úÖ **Validar formatos** (extens√µes, regex)  
‚úÖ **Validar tamanhos** (min, max)  
‚úÖ **Validar regras** de neg√≥cio simples  
‚úÖ **Retornar** `{'valid': bool, 'error': str, 'data': dict}`

‚ùå **NUNCA** acessar banco de dados  
‚ùå **NUNCA** fazer persist√™ncia  
‚ùå **NUNCA** fazer l√≥gica complexa

### Controllers (usando Validators)
```python
# controllers/file_upload_controller.py

from ..api.validators import FileUploadValidator
from ..repositories import FileUploadRepository

class FileUploadController:
    
    @staticmethod
    def create_upload(file_obj):
        """
        Create new file upload.
        
        Flow: Validate ‚Üí Repository ‚Üí Return
        """
        # 1. Validate
        validation = FileUploadValidator.validate_file(file_obj)
        
        if not validation['valid']:
            return {
                'success': False,
                'error': validation['error'],
                'upload': None
            }
        
        # 2. Repository
        try:
            upload = FileUploadRepository.create(
                file=file_obj,
                status='PENDING'
            )
        except Exception as e:
            return {
                'success': False,
                'error': f'Database error: {str(e)}',
                'upload': None
            }
        
        # 3. Return
        return {
            'success': True,
            'error': None,
            'upload': upload
        }
```

---

## üîç TIPOS DE VALIDA√á√ÉO

### 1. Valida√ß√£o de Formato
```python
@staticmethod
def validate_email(email):
    import re
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    if not re.match(pattern, email):
        return {'valid': False, 'error': 'Invalid email format'}
    return {'valid': True, 'error': None}
```

### 2. Valida√ß√£o de Range
```python
@staticmethod
def validate_year(year):
    current_year = datetime.now().year
    if year < 2020 or year > current_year + 1:
        return {'valid': False, 'error': f'Year must be between 2020 and {current_year + 1}'}
    return {'valid': True, 'error': None}
```

### 3. Valida√ß√£o de Lista
```python
@staticmethod
def validate_months(months):
    if not isinstance(months, list):
        return {'valid': False, 'error': 'Months must be a list'}
    
    for month in months:
        if month < 1 or month > 12:
            return {'valid': False, 'error': f'Invalid month: {month}'}
    
    return {'valid': True, 'error': None}
```

### 4. Valida√ß√£o de Depend√™ncias
```python
@staticmethod
def validate_date_range(start_date, end_date):
    if end_date < start_date:
        return {'valid': False, 'error': 'End date must be after start date'}
    
    max_range = timedelta(days=365)
    if (end_date - start_date) > max_range:
        return {'valid': False, 'error': 'Date range cannot exceed 1 year'}
    
    return {'valid': True, 'error': None}
```

---

## üö´ VIOLA√á√ïES COMUNS

### ‚ùå Exemplo 1: Valida√ß√£o no Controller
```python
# ‚ùå ERRADO
class FileUploadController:
    @staticmethod
    def create_upload(file_obj):
        # Valida√ß√£o inline ‚ùå
        if not file_obj:
            return {'success': False, 'error': 'No file'}
        
        if file_obj.size > 50*1024*1024:
            return {'success': False, 'error': 'Too large'}
        
        upload = FileUploadRepository.create(file=file_obj)
        return {'success': True, 'upload': upload}

# ‚úÖ CORRETO
class FileUploadController:
    @staticmethod
    def create_upload(file_obj):
        # Delega valida√ß√£o ‚úÖ
        validation = FileUploadValidator.validate_file(file_obj)
        
        if not validation['valid']:
            return {'success': False, 'error': validation['error']}
        
        upload = FileUploadRepository.create(file=file_obj)
        return {'success': True, 'upload': upload}
```

### ‚ùå Exemplo 2: Valida√ß√£o na View
```python
# ‚ùå ERRADO
class UploadAPIView(APIView):
    def post(self, request):
        file = request.FILES.get('file')
        
        # Valida√ß√£o na view ‚ùå
        if not file.name.endswith('.xlsx'):
            return Response({'error': 'Invalid'}, status=400)
        
        result = FileUploadController.create_upload(file)
        return Response(result)

# ‚úÖ CORRETO
class UploadAPIView(APIView):
    def post(self, request):
        file = request.FILES.get('file')
        
        # Controller faz valida√ß√£o via Validator ‚úÖ
        result = FileUploadController.create_upload(file)
        
        if not result['success']:
            return Response({'error': result['error']}, status=400)
        
        serializer = FileUploadSerializer(result['upload'])
        return Response(serializer.data, status=201)
```

---

## ‚úÖ CHECKLIST

### Estrutura
- [ ] `api/validators.py` existe
- [ ] Cada model tem seu Validator
- [ ] Controllers usam Validators
- [ ] Views n√£o fazem valida√ß√£o

### Validator
- [ ] M√©todos est√°ticos
- [ ] Retorna dict `{'valid', 'error', 'data'}`
- [ ] N√£o acessa banco
- [ ] N√£o faz persist√™ncia
- [ ] Apenas valida dados

### Controller
- [ ] Chama Validator primeiro
- [ ] Trata resultado de valida√ß√£o
- [ ] Retorna dict padronizado
- [ ] N√£o faz valida√ß√£o inline

---

**VALIDATORS = INPUT VALIDATION. CONTROLLERS = ORCHESTRATION.**

